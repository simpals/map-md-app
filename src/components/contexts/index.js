import { createContext } from 'react';

export const MapCtx = createContext({
  MapGL: null,
  userLocation: null,
  authorization: null,
  style: null,
  changeStyle: () => {}
});
