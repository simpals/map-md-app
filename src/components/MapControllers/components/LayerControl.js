import React, { useState, useContext } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { MapCtx } from '../../contexts';
import { MAP_STYLE } from '../../../constants';
import objects3DRender from '../../../Render3D';

import viewTypeIcons from '../../../static/images/view-type-icons.png';

const Wrap = styled.div`
  position: absolute;
  right: 15px;
  bottom: 13px;
  z-index: 100;
  ${({ theme: { layerSwitcher } }) => layerSwitcher && layerSwitcher}
`;

const Switcher = styled.a`
  cursor: pointer;
  display: inline-block;
  width: 55px;
  height: 55px;
  margin-left: 10px;
  border: 3px solid #fff;
  border-radius: 50%;
  background-image: url(${viewTypeIcons});
  background-position: ${({ layer }) => (layer === 'satelite' ? '0px' : '-49px')} 0px;
`;

const LayerControl = ({ disable3DSatelite, disable3DMode, mapImages }) => {
  const { OPTIONS, URL } = MAP_STYLE;
  const {
    MapGL, authorization, style, changeStyle
  } = useContext(MapCtx);
  const [layer, setLayer] = useState(style);
  OPTIONS.headers.authorization = authorization;

  const mapGetData = async (nextLayer) => {
    const styleReq = await fetch(URL[nextLayer], OPTIONS);
    const mapStyles = await styleReq.json();
    const initialSource = MapGL.getStyle().sources;
    MapGL.getStyle()
      .layers.slice(0)
      .reduceRight((acc, mapLayer, _, originArr) => {
        if (!mapLayer.id.match(/(vector-|raster-)/)) {
          acc.push(mapLayer);
        } else {
          mapStyles.layers.push(...acc.reverse());
          originArr.splice(0);
        }
        return acc;
      }, []);
    Object.keys(initialSource).map((source) => {
      if (initialSource[source].type !== 'vector' && initialSource[source].type !== 'raster') {
        mapStyles.sources[source] = initialSource[source];
      }
      return null;
    });
    return Promise.resolve(mapStyles);
  };

  const changeLayer = async (currentLayer, nextLayer) => {
    if (currentLayer === nextLayer) return;
    const mapData = await mapGetData(nextLayer);
    MapGL.setStyle(mapData, { diff: false });
    mapImages.map((image) => {
      const { id, path } = image;
      MapGL.loadImage(path, (_, imageFile) => {
        if (!MapGL.hasImage(id) && imageFile) MapGL.addImage(id, imageFile);
      });
    });
    MapGL.once('styledata', () => {
      if (MapGL.getPitch() <= 0 || (nextLayer === 'satelite' && disable3DSatelite)) return;
      if (!disable3DMode) objects3DRender(MapGL);
    });
    setLayer(nextLayer);
    changeStyle(nextLayer);
  };

  const nextLayer = layer === 'satelite' ? 'map' : 'satelite';

  return (
    <Wrap className='LayerControl' >
      <Switcher onClick={() => changeLayer(layer, nextLayer)} layer={layer} />
    </Wrap>
  );
};

LayerControl.propTypes = {
  disable3DSatelite: PropTypes.bool.isRequired,
  disable3DMode: PropTypes.bool.isRequired,
  mapImages: PropTypes.array.isRequired
};

export default LayerControl;
