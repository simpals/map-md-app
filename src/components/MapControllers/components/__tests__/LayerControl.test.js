import 'jsdom-global/register';
import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import LayerControl from '../LayerControl';
import { MapCtx } from '../../../contexts';

const setApiKey = ''; // API key
const ENCODE = Buffer.from(`${setApiKey}:pashalka`).toString('base64');
const AUTHORIZATION = `Basic ${ENCODE}`;

const context = {
  MapGL: {

  },
  authorization: AUTHORIZATION,
  style: 'map',
  changeStyle: jest.fn()
};

const props = {
  disable3DSatelite: false,
  disable3DMode: false,
  mapImages: []
};

describe('test LayerControl component', () => {
  it('test LayerControl mount', (done) => {
    const component = mount(
      <MapCtx.Provider value={context}>
        <LayerControl {...props} />
      </MapCtx.Provider>
    );
    if (!component) done.fail(new Error('Mount fail'));
    done();
  });

  it('test LayerControl matchSnapshot', () => {
    const component = shallow(<LayerControl {...props} />);
    expect(toJson(component)).toMatchSnapshot();
  });
});
