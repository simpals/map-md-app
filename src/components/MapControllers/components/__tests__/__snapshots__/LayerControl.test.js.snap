// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test LayerControl component test LayerControl matchSnapshot 1`] = `
<styled.div
  className="LayerControl"
>
  <styled.a
    layer={null}
    onClick={[Function]}
  />
</styled.div>
`;
